src/main/scala/org/bitcoins/core/util/BitcoinScriptUtil.scala

```diff
@@ -297,23 +231,156 @@ trait BitcoinScriptUtil {
+          //shouldn't have BaseTransactionSignatureComponent with a witness scriptPubKey
+          script
+      }
+    case  _ : P2PKHScriptPubKey | _ : P2PKScriptPubKey | _ : MultiSignatureScriptPubKey |
+         _ : NonStandardScriptPubKey | _ : CLTVScriptPubKey | _ : CSVScriptPubKey | _ : WitnessCommitment | EmptyScriptPubKey =>
+      script
+  }
+
+  /** Removes the given [[ECDigitalSignature]] from the list of [[ScriptToken]] if it exists. */
+  def removeSignatureFromScript(signature : ECDigitalSignature, script : Seq[ScriptToken]) : Seq[ScriptToken] = {
+    if (script.contains(ScriptConstant(signature.hex))) {
+      //replicates this line in bitcoin core
+      //https://github.com/bitcoin/bitcoin/blob/master/src/script/interpreter.cpp#L872
+      val sigIndex = script.indexOf(ScriptConstant(signature.hex))
+      logger.debug("SigIndex: " + sigIndex)
+      //remove sig and it's corresponding BytesToPushOntoStack
+      val sigRemoved = script.slice(0,sigIndex-1) ++ script.slice(sigIndex+1,script.size)
+      logger.debug("sigRemoved: " + sigRemoved)
+      sigRemoved
+    } else script
+  }
+
+  /** Removes the list of [[ECDigitalSignature]] from the list of [[ScriptToken]] */
+  def removeSignaturesFromScript(sigs : Seq[ECDigitalSignature], script : Seq[ScriptToken]) : Seq[ScriptToken] = {
+    @tailrec
+    def loop(remainingSigs : Seq[ECDigitalSignature], scriptTokens : Seq[ScriptToken]) : Seq[ScriptToken] = {
+      remainingSigs match {
+        case Nil => scriptTokens
+        case h :: t =>
+          val newScriptTokens = removeSignatureFromScript(h,scriptTokens)
+          loop(t,newScriptTokens)
+      }
+    }
+    loop(sigs,script)
+  }
+
+  /** Removes the [[org.bitcoins.core.script.crypto.OP_CODESEPARATOR]] in the original script according to
+    * the last code separator index in the script. */
+  def removeOpCodeSeparator(program : ExecutionInProgressScriptProgram) : Seq[ScriptToken] = {
+    if (program.lastCodeSeparator.isDefined) {
+      program.originalScript.slice(program.lastCodeSeparator.get+1, program.originalScript.size)
+    } else program.originalScript
+  }
+
+  private def parseScriptEither(scriptEither: Either[(Seq[ScriptToken], ScriptPubKey), ScriptError]): Seq[ScriptToken] = scriptEither match {
+    case Left((_,scriptPubKey)) =>
+      logger.debug("Script pubkey asm inside calculateForSigning: " + scriptPubKey.asm)
+      scriptPubKey.asm
+    case Right(_) => Nil //error
+  }
+
+  /** Given a tx, scriptPubKey and the input index we are checking the tx, it derives the appropriate [[SignatureVersion]] to use */
+  @tailrec
+  final def parseSigVersion(tx: Transaction, scriptPubKey: ScriptPubKey, inputIndex: UInt32): SignatureVersion  = scriptPubKey match {
+    case _ : WitnessScriptPubKeyV0 | _: UnassignedWitnessScriptPubKey =>
+      SigVersionWitnessV0
+    case _: P2SHScriptPubKey =>
+      val scriptSig = tx.inputs(inputIndex.toInt).scriptSignature
```

Comment:
Since every `P2SHScriptPubKey` has a corresponding `P2SHScriptSignature` we can remove this pattern matching here and just assume that the scriptSig is a `P2SHScriptSignature`